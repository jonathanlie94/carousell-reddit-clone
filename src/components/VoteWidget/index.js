/*
  A widget that displays the upvote button, vote count, and downvote button.
*/

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { theme } from 'ui';
import { noselect } from 'ui/helpers';

const Wrapper = styled.div`
  ${noselect()} display: inline-flex;
  flex-direction: column;
  text-align: center;
  min-width: 3.25rem;
`;

const ArrowIcon = styled.i`
  color: ${theme.colors.grey};
  /* We use pixel for the fontSize here so that the size does not change
    with the browser's font size. */
  font-size: 32px !important;
  cursor: pointer;
  &:hover {
    opacity: 0.5;
  }
`;

const ZeroCountIcon = styled.i`
  /* Margin to account for font size of normal text */
  font-size: 12px !important;
  margin: 3px 0;
  color: ${theme.colors.grey};
`;

class VoteWidget extends Component {
  render() {
    return (
      <Wrapper>
        <ArrowIcon
          className="material-icons"
          onClick={this.props.onUpvote}
          ref="upvote"
        >
          arrow_drop_up
        </ArrowIcon>
        {this.props.count !== 0
          ? <div>
              {this.props.count}
            </div>
          : <ZeroCountIcon className="material-icons">lens</ZeroCountIcon>}
        <ArrowIcon
          className="material-icons"
          onClick={this.props.onDownvote}
          ref="downvote"
        >
          arrow_drop_down
        </ArrowIcon>
      </Wrapper>
    );
  }
}

VoteWidget.propTypes = {
  onUpvote: PropTypes.func,
  onDownvote: PropTypes.func,
  count: PropTypes.number,
};

export default VoteWidget;
