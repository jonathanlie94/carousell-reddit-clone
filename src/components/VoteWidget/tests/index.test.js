import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import 'jest-styled-components';
import VoteWidget from '../';

describe('<VoteWidget />', () => {
  it('should handle upvote correctly', () => {
    let count = 2;
    const onUpvote = v => {
      count++;
    };
    const renderedComponent = mount(
      <VoteWidget onUpvote={onUpvote} count={count} />
    );
    renderedComponent.ref('upvote').simulate('click');

    expect(count).toEqual(3);
  });

  it('should handle downvote correctly', () => {
    let count = 2;
    const onDownvote = v => {
      count--;
    };
    const renderedComponent = mount(
      <VoteWidget onDownvote={onDownvote} count={count} />
    );
    renderedComponent.ref('downvote').simulate('click');

    expect(count).toEqual(1);
  });

  it('should match previous snapshot', () => {
    const zeroComponent = mount(<VoteWidget count={0} />);
    const zero = toJSON(zeroComponent);
    expect(zero).toMatchStyledComponentsSnapshot();

    const positiveComponent = mount(<VoteWidget count={1} />);
    const positive = toJSON(positiveComponent);
    expect(positive).toMatchStyledComponentsSnapshot();

    const negativeComponent = mount(<VoteWidget count={-1} />);
    const negative = toJSON(negativeComponent);
    expect(negative).toMatchStyledComponentsSnapshot();
  });
});
