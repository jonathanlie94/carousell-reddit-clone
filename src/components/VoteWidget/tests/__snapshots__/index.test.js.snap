// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VoteWidget /> should match previous snapshot 1`] = `
.gAjEpo {
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: center;
  min-width: 3.25rem;
}

.fnsJyR {
  color: #C8C8C8;
  font-size: 32px !important;
  cursor: pointer;
}

.fnsJyR:hover {
  opacity: 0.5;
}

.heBpaD {
  font-size: 12px !important;
  margin: 3px 0;
  color: #C8C8C8;
}

<VoteWidget
  count={0}
>
  <styled.div>
    <div
      className="sc-bdVaJa gAjEpo"
    >
      <styled.i
        className="material-icons"
      >
        <i
          className="material-icons sc-bwzfXH fnsJyR"
        >
          arrow_drop_up
        </i>
      </styled.i>
      <styled.i
        className="material-icons"
      >
        <i
          className="material-icons sc-htpNat heBpaD"
        >
          lens
        </i>
      </styled.i>
      <styled.i
        className="material-icons"
      >
        <i
          className="material-icons sc-bwzfXH fnsJyR"
        >
          arrow_drop_down
        </i>
      </styled.i>
    </div>
  </styled.div>
</VoteWidget>
`;

exports[`<VoteWidget /> should match previous snapshot 2`] = `
.gAjEpo {
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: center;
  min-width: 3.25rem;
}

.fnsJyR {
  color: #C8C8C8;
  font-size: 32px !important;
  cursor: pointer;
}

.fnsJyR:hover {
  opacity: 0.5;
}

<VoteWidget
  count={1}
>
  <styled.div>
    <div
      className="sc-bdVaJa gAjEpo"
    >
      <styled.i
        className="material-icons"
      >
        <i
          className="material-icons sc-bwzfXH fnsJyR"
        >
          arrow_drop_up
        </i>
      </styled.i>
      <div>
        1
      </div>
      <styled.i
        className="material-icons"
      >
        <i
          className="material-icons sc-bwzfXH fnsJyR"
        >
          arrow_drop_down
        </i>
      </styled.i>
    </div>
  </styled.div>
</VoteWidget>
`;

exports[`<VoteWidget /> should match previous snapshot 3`] = `
.gAjEpo {
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: center;
  min-width: 3.25rem;
}

.fnsJyR {
  color: #C8C8C8;
  font-size: 32px !important;
  cursor: pointer;
}

.fnsJyR:hover {
  opacity: 0.5;
}

<VoteWidget
  count={-1}
>
  <styled.div>
    <div
      className="sc-bdVaJa gAjEpo"
    >
      <styled.i
        className="material-icons"
      >
        <i
          className="material-icons sc-bwzfXH fnsJyR"
        >
          arrow_drop_up
        </i>
      </styled.i>
      <div>
        -1
      </div>
      <styled.i
        className="material-icons"
      >
        <i
          className="material-icons sc-bwzfXH fnsJyR"
        >
          arrow_drop_down
        </i>
      </styled.i>
    </div>
  </styled.div>
</VoteWidget>
`;
